src/cart/src/cart/crt_group.c

```diff
@@ -359,41 +335,39 @@ grp_li_uri_set(struct crt_lookup_item *li, int tag, const char *uri)
 		D_INIT_LIST_HEAD(&ui->ui_link);
 		ui->ui_ref = 0;
 		ui->ui_initialized = 1;
-
-		rc = D_MUTEX_INIT(&ui->ui_mutex, NULL);
-		if (rc != 0) {
+		ui->ui_rank = li->li_rank;
+		D_STRNDUP(ui->ui_uri[tag], uri, CRT_ADDR_STR_MAX_LEN);
+		if (!ui->ui_uri[tag]) {
+			D_ERROR("Failed to duplicate an uri\n");
 			D_FREE_PTR(ui);
-			D_GOTO(exit, rc);
+			D_GOTO(exit, rc = -DER_NOMEM);
 		}
 
-		ui->ui_rank = li->li_rank;
-
 		rc = d_hash_rec_insert(&grp_priv->gp_uri_lookup_cache,
 				&rank, sizeof(rank),
 				&ui->ui_link,
 				true /* exclusive */);
 		if (rc != 0) {
 			D_ERROR("Entry already present\n");
-			D_MUTEX_DESTROY(&ui->ui_mutex);
+			D_FREE(ui->ui_uri[tag]);
 			D_FREE_PTR(ui);
 			D_GOTO(exit, rc);
 		}
-		D_STRNDUP(ui->ui_uri[tag], uri, CRT_ADDR_STR_MAX_LEN);
-
-		if (!ui->ui_uri[tag]) {
-			d_hash_rec_delete(&grp_priv->gp_uri_lookup_cache,
-					&rank, sizeof(d_rank_t));
-			D_GOTO(exit, rc = -DER_NOMEM);
-		}
 	} else {
 		ui = crt_ui_link2ptr(rlink);
-		if (!ui->ui_uri[tag]) {
-			D_STRNDUP(ui->ui_uri[tag], uri, CRT_ADDR_STR_MAX_LEN);
-		}
+		if (ui->ui_uri[tag] == NULL) {
```

Comment:
This still isn't right is it?
I think you need to always allocate the string, then try the atomic_compare_exchange, and if it fails the compare then free the string again.  It also looks like you're not modifying rc if the == NULL check fails, but setting it to -DER_ALREADY if the compare_exchange fails, and I suspect these two should be consistent.

Reply:
This is just fast double check to avoid string duplication. This potentially can save long duplication operation.
After duplication we exchange new string with old value and if it stay NULL, like in previous check, we are done. Otherwise, somebody set the value already, we free just duplicated string and return -EALREADY.

Reply:
I was thinking that the cost of the string duplication was something that we had to accept for the benefit of using atomics, but I can see with the double-check trick that the incidence of having to needlessly allocate the string is much reduced.

If you're doing this shouldn't we return -DER_ALREADY if it's already present and the first check fails?

I can try and find the documentation relating to the matter but I think if we're using the atomic interface then there's a requirement that variables are *only* accessed through the atomic functions, which would me that you needed to add back in the atomic wrapper on the first call.  Sorry for mis-describing the problem.

Reply:
Ohh. Unfortunately the error code -DER_ALREADY is not handled properly. So, it would be better to stay the logic the same and keep silent about this collision.

Hmm. I don't think we have this strict requirements to use _only_ atomic interface... I have to clarify this in documentation.

Reply:
Quote from documentation: 

> Built-in increment and decrement operators and compound assignment are read-modify-
> write atomic operations with total sequentially consistent ordering (as if using
> memory_order_seq_cst). **If less strict synchronization semantics are desired, the standard
> library functions may be used instead**.